packages/gatsby/src/schema/node-model.js

```diff
@@ -278,6 +485,251 @@ const toNodeTypeNames = (schema, gqlTypeName) => {
+    const v = value[k]
+    if (_.isPlainObject(value) && _.isPlainObject(v)) {
+      acc[key] =
+        k === `elemMatch` ? dropQueryOperators(v) : dropQueryOperators(value)
+    } else {
+      acc[key] = true
+    }
+    return acc
+  }, {})
+
+const mergeObjects = (obj1, obj2) =>
+  Object.keys(obj2).reduce((acc, key) => {
+    const value = obj2[key]
+    if (typeof value === `object` && value && acc[key]) {
+      acc[key] = mergeObjects(acc[key], value)
+    } else {
+      acc[key] = value
+    }
+    return acc
+  }, obj1)
+
+const merge = (...objects) => {
+  const [first, ...rest] = objects.filter(Boolean)
+  return rest.reduce((acc, obj) => mergeObjects(acc, obj), { ...first })
+}
+
+async function resolveRecursive(
+  nodeModel,
+  schemaComposer,
+  schema,
+  node,
+  type,
+  queryFields,
+  fieldsToResolve
+) {
+  const gqlFields = type.getFields()
+  let resolvedFields = {}
+  for (const fieldName of Object.keys(fieldsToResolve)) {
+    const fieldToResolve = fieldsToResolve[fieldName]
+    const queryField = queryFields[fieldName]
+    const gqlField = gqlFields[fieldName]
+    const gqlNonNullType = getNullableType(gqlField.type)
+    const gqlFieldType = getNamedType(gqlField.type)
+    let innerValue
+    if (gqlField.resolve) {
+      innerValue = await resolveField(
+        nodeModel,
+        schemaComposer,
+        schema,
+        node,
+        gqlField,
+        fieldName
+      )
+    } else {
+      innerValue = node[fieldName]
+    }
+    if (gqlField && innerValue != null) {
+      if (
+        isCompositeType(gqlFieldType) &&
+        !(gqlNonNullType instanceof GraphQLList)
+      ) {
+        innerValue = await resolveRecursive(
+          nodeModel,
+          schemaComposer,
+          schema,
+          innerValue,
+          gqlFieldType,
+          queryField,
+          _.isObject(fieldToResolve) ? fieldToResolve : queryField
+        )
+      } else if (
+        isCompositeType(gqlFieldType) &&
+        _.isArray(innerValue) &&
+        gqlNonNullType instanceof GraphQLList
+      ) {
+        innerValue = await Promise.all(
+          innerValue.map(item =>
+            resolveRecursive(
+              nodeModel,
+              schemaComposer,
+              schema,
+              item,
+              gqlFieldType,
+              queryField,
+              _.isObject(fieldToResolve) ? fieldToResolve : queryField
+            )
+          )
+        )
+      }
+    }
+    if (innerValue != null) {
+      resolvedFields[fieldName] = innerValue
+    }
+  }
+
+  Object.keys(queryFields).forEach(key => {
+    if (!fieldsToResolve[key] && node[key]) {
+      resolvedFields[key] = node[key]
+    }
+  })
+
+  return _.pickBy(resolvedFields, (value, key) => queryFields[key])
+}
+
+function resolveField(
+  nodeModel,
+  schemaComposer,
+  schema,
+  node,
+  gqlField,
+  fieldName
+) {
+  const withResolverContext = require(`./context`)
+  return gqlField.resolve(
+    node,
+    {},
```

Comment:
we should pass the args defaultValues here, something like
```suggestion
    gqlField.args.reduce((acc, arg) => {
      acc[arg.name] = arg.defaultValue
      return acc
    }, {}),
```